### PR TITLE
Add a node for the SCS branch and bound

### PR DESCRIPTION
**Pull request by hongkai-dai**
This is the first PR to write our own branch and bound (bnb) for mixed-integer optimization. In this PR I add a class `ScsNode`, as a node in the tree for the branch and bound algorithm, used by SCS solver. Since this PR already has > 1000 lines of change, I do not include the implementation on the branch and bound algorithm in this PR. The bnb implementation will be included in the next PR.

cc @RussTedrake

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/7526)

### CODE REVIEW

**Comment by hongkai-dai** [comment]
+@avalenzu for feature review please.

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7526#-:-L-0RSAIr2iqRytVF8aZ:bll7v0f)*

**Comment by hongkai-dai** [comment]
+@robert-verkuil for feature review please.

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7526#-:-L-3LJmTbOZt70ohsW5m:b-vr56ki)*

**Comment by avalenzu** [comment]
Checkpoint.

---

Reviewed 1 of 2 files at r2, 1 of 1 files at r3.
Review status: 2 of 4 files reviewed at latest revision, 13 unresolved discussions.

---

*[drake/solvers/BUILD.bazel, line 1090 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7526#-L-dNXKedQ0Drxmd0OYG:-L-dNXKedQ0Drxmd0OYH:bry73ao) ([raw file](https://github.com/robotlocomotion/drake/blob/a16c56fb091ae1913480c89599cdabbbadb595b6/drake/solvers/BUILD.bazel#L1090)):*
> ```
>     "no_snopt.cc",
>     "qp.cc",
>     "scs_bnb.cc",
> ```

I know that this filename name matches `scs` but it seems too short to me.

---

*[drake/solvers/scs_bnb.cc, line 7 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7526#-L-WxGQ2z_vN1_jZ2zxd:-L-WxGQ2z_vN1_jZ2zxe:b-x3r76j) ([raw file](https://github.com/robotlocomotion/drake/blob/81394807e9cf1c222c89d3089e53d2ebd6d2e8f5/drake/solvers/scs_bnb.cc#L7)):*
> ```cpp
> #include <unordered_set>
> 
> // This code is in C style, to be compatible with the SCS source code.
> ```

What exactly do you mean by this? I see a decidedly un-C-like `std::unique_ptr` a few lines below.

---

*[drake/solvers/scs_bnb.h, line 6 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7526#-L-WxiM8gbAJFIXwdba_:-L-WxiM8gbAJFIXwdbaa:b-mj3j9e) ([raw file](https://github.com/robotlocomotion/drake/blob/81394807e9cf1c222c89d3089e53d2ebd6d2e8f5/drake/solvers/scs_bnb.h#L6)):*
> ```objc
> #include <memory>
> 
> // clang-format off
> ```

Could you add a note as to why this is necessary?

---

*[drake/solvers/scs_bnb.h, line 66 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7526#-L-gFF5wlUcmr0xQ5kkP:-L-gFF5wlUcmr0xQ5kkQ:b-qshinf) ([raw file](https://github.com/robotlocomotion/drake/blob/a16c56fb091ae1913480c89599cdabbbadb595b6/drake/solvers/scs_bnb.h#L66)):*
> The life of the cone must outlive the life of the node and the tree for the branch and bound.

The phrasing we've used for this elsewhere (e.g. [direct_collocation.h](https://github.com/RobotLocomotion/drake/blob/627914db8df950126aedbf723ab81f41f06c989a/drake/systems/trajectory_optimization/direct_collocation.h#L32)) is

> Note that this is aliased for the lifetime of this object.

Would that work here?

---

*[drake/solvers/scs_bnb.h, line 67 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7526#-L-gG45L3Xp8dph3sG9c:-L-gG45M9o6zqqPVsK41:b-5g43oh) ([raw file](https://github.com/robotlocomotion/drake/blob/a16c56fb091ae1913480c89599cdabbbadb595b6/drake/solvers/scs_bnb.h#L67)):*
I suggest making this a little more explicit:

> The indices of the binary variables, y, in the vector of decision variables, x.

---

*[drake/solvers/scs_bnb.h, line 90 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7526#-L-gGtPob8paan6Sh5oi:-L-gGtPob8paan6Sh5oj:b-z7j06o) ([raw file](https://github.com/robotlocomotion/drake/blob/a16c56fb091ae1913480c89599cdabbbadb595b6/drake/solvers/scs_bnb.h#L90)):*
> ```objc
>    * @throw std::runtime_error if the preconditions are not met.
>    */
>   void Branch(int binary_var_index);
> ```

Can `Branch()` be called more than once on a given `ScsNode`?

---

*[drake/solvers/scs_bnb.h, line 104 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7526#-L-gHpVsA5Wj71DS0rWs:-L-gHpVsA5Wj71DS0rWt:bgsz4ru) ([raw file](https://github.com/robotlocomotion/drake/blob/a16c56fb091ae1913480c89599cdabbbadb595b6/drake/solvers/scs_bnb.h#L104)):*
> ```objc
>    * 2. The problem is feasible, we can then update the lower bound, as the
>    *    minimal among all the costs in the leaf nodes.
>    * 3. When the problem is feasible, and we find a solution that satisfies the
> ```

"When the problem..." -> "The problem"

---

*[drake/solvers/scs_bnb.h, line 106 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7526#-L-gI2Jg2J6zB8yQ_ecd:-L-gI2Jg2J6zB8yQ_ece:b-y8qysm) ([raw file](https://github.com/robotlocomotion/drake/blob/a16c56fb091ae1913480c89599cdabbbadb595b6/drake/solvers/scs_bnb.h#L106)):*
> If the cost of this solution is an upper bound of the original mixed-integer problem.

Should this just be "The cost of this solution ..."?

---

*[drake/solvers/scs_bnb.h, line 108 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7526#-L-gII459MwGJKmDXvvp:-L-gII459MwGJKmDXvvq:b-ygw3z2) ([raw file](https://github.com/robotlocomotion/drake/blob/a16c56fb091ae1913480c89599cdabbbadb595b6/drake/solvers/scs_bnb.h#L108)):*
Consider replacing this with

> The problem is feasible, but the optimal cost is larger than the best upper bound. Then there is no need to branch on this node.

---

*[drake/solvers/scs_bnb.h, line 138 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7526#-L-gP75dJPtJFfpQeZ8x:-L-gP75dJPtJFfpQeZ8y:b-ssoolr) ([raw file](https://github.com/robotlocomotion/drake/blob/a16c56fb091ae1913480c89599cdabbbadb595b6/drake/solvers/scs_bnb.h#L138)):*
> ```objc
> 
>   /**
>    * This node is created from its parent node, by branching on a binary
> ```

"is" -> "was"

---

*[drake/solvers/scs_bnb.h, line 144 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7526#-L-gPA_lZ8cK3JJrpkXO:-L-gPA_lZ8cK3JJrpkXP:b-ssoolr) ([raw file](https://github.com/robotlocomotion/drake/blob/a16c56fb091ae1913480c89599cdabbbadb595b6/drake/solvers/scs_bnb.h#L144)):*
> ```objc
> 
>   /**
>    * This node is created from its parent node, by fixing a binary variable to
> ```

"is" -> "was"

---

*[drake/solvers/scs_bnb.h, line 195 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7526#-L-gPgGmU2qxGPokvTkF:-L-gPgGmU2qxGPokvTkG:b-leylgg) ([raw file](https://github.com/robotlocomotion/drake/blob/a16c56fb091ae1913480c89599cdabbbadb595b6/drake/solvers/scs_bnb.h#L195)):*
> ```objc
>   double cost_;
>   // If a node is fathomed, then there is no need to branch on this node.
>   // There are two possible cases to make a node fathomed_
> ```

Stray underscore in "fathomed_"

---

*[drake/solvers/scs_bnb.h, line 200 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7526#-L-gPmDToGG91Ku1zsCX:-L-gPmDToGG91Ku1zsCY:b8vjqkd) ([raw file](https://github.com/robotlocomotion/drake/blob/a16c56fb091ae1913480c89599cdabbbadb595b6/drake/solvers/scs_bnb.h#L200)):*
> ```objc
>   // 2. The optimal solution to the relaxed problem is larger than the current
>   //    best upper bound.
>   bool found_integral_sol_;
> ```

If `found_integral_sol_` is false, is there a way to tell that the node is fathomed due to (2)?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7526#-:-L-gPx_fmPqrk2NJO0k7:by96gdd)*

**Comment by hongkai-dai** [comment]
Review status: 2 of 4 files reviewed at latest revision, 13 unresolved discussions.

---

*[drake/solvers/BUILD.bazel, line 1090 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7526#-L-dNXKedQ0Drxmd0OYG:-L-jLJBWGTaIx6eg243V:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a16c56fb091ae1913480c89599cdabbbadb595b6/drake/solvers/BUILD.bazel#L1090)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

I know that this filename name matches `scs` but it seems too short to me.
</blockquote></details>

Done.

---

*[drake/solvers/scs_bnb.cc, line 7 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7526#-L-WxGQ2z_vN1_jZ2zxd:-L-jLLDVvU4QXER59463:b91htoi) ([raw file](https://github.com/robotlocomotion/drake/blob/81394807e9cf1c222c89d3089e53d2ebd6d2e8f5/drake/solvers/scs_bnb.cc#L7)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

What exactly do you mean by this? I see a decidedly un-C-like `std::unique_ptr` a few lines below.
</blockquote></details>

I mean I have input like `const scs_float* const b`, which is a raw pointer.

---

*[drake/solvers/scs_bnb.h, line 6 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7526#-L-WxiM8gbAJFIXwdba_:-L-jLy_zmBFku-kmsj6Y:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/81394807e9cf1c222c89d3089e53d2ebd6d2e8f5/drake/solvers/scs_bnb.h#L6)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

Could you add a note as to why this is necessary?
</blockquote></details>

Done.

---

*[drake/solvers/scs_bnb.h, line 66 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7526#-L-gFF5wlUcmr0xQ5kkP:-L-jMW53mtqj1jkV8dti:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a16c56fb091ae1913480c89599cdabbbadb595b6/drake/solvers/scs_bnb.h#L66)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

> The life of the cone must outlive the life of the node and the tree for the branch and bound.

The phrasing we've used for this elsewhere (e.g. [direct_collocation.h](https://github.com/RobotLocomotion/drake/blob/627914db8df950126aedbf723ab81f41f06c989a/drake/systems/trajectory_optimization/direct_collocation.h#L32)) is

> Note that this is aliased for the lifetime of this object.

Would that work here?

</blockquote></details>

Done.

---

*[drake/solvers/scs_bnb.h, line 67 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7526#-L-gG45L3Xp8dph3sG9c:-L-jMbbnrHnA-HBOXIFd:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a16c56fb091ae1913480c89599cdabbbadb595b6/drake/solvers/scs_bnb.h#L67)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

I suggest making this a little more explicit:

> The indices of the binary variables, y, in the vector of decision variables, x.
</blockquote></details>

Done.

---

*[drake/solvers/scs_bnb.h, line 90 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7526#-L-gGtPob8paan6Sh5oi:-L-jMcV7UOBtEBAkcR3-:b-6mjnb5) ([raw file](https://github.com/robotlocomotion/drake/blob/a16c56fb091ae1913480c89599cdabbbadb595b6/drake/solvers/scs_bnb.h#L90)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

Can `Branch()` be called more than once on a given `ScsNode`?
</blockquote></details>

Excellent question! When we do branch and bound, in order to quickly find an integral solution, one approach is to fix all binary variables in one node, and solve for the continuous variables. So in this approach, essentially it is trying to fix the binary variables for multiple times, which is what `Branch` does.

Another case is that when we want to warm-start a mixed-integer problem, by going up from the leaf of the tree, we might want to branch on a node again, that has been branched, due to the slight change on the problem data.

---

*[drake/solvers/scs_bnb.h, line 104 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7526#-L-gHpVsA5Wj71DS0rWs:-L-jQEy6vybhByqrYif7:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a16c56fb091ae1913480c89599cdabbbadb595b6/drake/solvers/scs_bnb.h#L104)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

"When the problem..." -> "The problem"
</blockquote></details>

Done.

---

*[drake/solvers/scs_bnb.h, line 106 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7526#-L-gI2Jg2J6zB8yQ_ecd:-L-jQOLqKWdqczSWVtZ7:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a16c56fb091ae1913480c89599cdabbbadb595b6/drake/solvers/scs_bnb.h#L106)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

> If the cost of this solution is an upper bound of the original mixed-integer problem.

Should this just be "The cost of this solution ..."?
</blockquote></details>

Done.

---

*[drake/solvers/scs_bnb.h, line 108 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7526#-L-gII459MwGJKmDXvvp:-L-jQWlbmrTqxY5iw5gN:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a16c56fb091ae1913480c89599cdabbbadb595b6/drake/solvers/scs_bnb.h#L108)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

Consider replacing this with

> The problem is feasible, but the optimal cost is larger than the best upper bound. Then there is no need to branch on this node.
</blockquote></details>

Done.

---

*[drake/solvers/scs_bnb.h, line 138 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7526#-L-gP75dJPtJFfpQeZ8x:-L-jQ_J5hcKBA9WzSjyc:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a16c56fb091ae1913480c89599cdabbbadb595b6/drake/solvers/scs_bnb.h#L138)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

"is" -> "was"
</blockquote></details>

Done.

---

*[drake/solvers/scs_bnb.h, line 144 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7526#-L-gPA_lZ8cK3JJrpkXO:-L-jQayC3SEgJoDby42R:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a16c56fb091ae1913480c89599cdabbbadb595b6/drake/solvers/scs_bnb.h#L144)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

"is" -> "was"
</blockquote></details>

Done.

---

*[drake/solvers/scs_bnb.h, line 195 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7526#-L-gPgGmU2qxGPokvTkF:-L-jQjyTje7AcrQGSHPs:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a16c56fb091ae1913480c89599cdabbbadb595b6/drake/solvers/scs_bnb.h#L195)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

Stray underscore in "fathomed_"
</blockquote></details>

Done.

---

*[drake/solvers/scs_bnb.h, line 200 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7526#-L-gPmDToGG91Ku1zsCX:-L-jQlQzs9y0WOz8d1nF:bdg9v4u) ([raw file](https://github.com/robotlocomotion/drake/blob/a16c56fb091ae1913480c89599cdabbbadb595b6/drake/solvers/scs_bnb.h#L200)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

If `found_integral_sol_` is false, is there a way to tell that the node is fathomed due to (2)?
</blockquote></details>

Yes, we will tell it in the tree, where the best upper bound is stored. This is implemented in my dev branch https://github.com/hongkai-dai/drake/blob/scs_bnb_tree/drake/solvers/scs_bnb.h#L345-L351, and will be included in my next PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7526)*

**Comment by avalenzu** [comment]
Meta-question: Have you considered writing C++ wrappers for the main SCS classes you use here? SCS's single letter variable names make this PR really hard to parse (for me at least). It might be preferable to isolate the C-ness in a wrapper layer, rather than having it combined with the actual branch-and-bound solver. What do you think?

---

Reviewed 3 of 7 files at r4.
Review status: 2 of 4 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[drake/solvers/scs_branch_and_bound.cc, line 62 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7526#-L-t0nTSj630DkQg9LOv:-L-t0nTSj630DkQg9LOw:b-f3shkp) ([raw file](https://github.com/robotlocomotion/drake/blob/f5b53283ec0988ff6b1435d6cfa0d4921ada56c5/drake/solvers/scs_branch_and_bound.cc#L62)):*
> ```cpp
>   root->binary_var_indices_ = binary_var_indices;
>   root->cost_constant_ = cost_constant;
>   // We need to add the constraint 0 ≤ y ≤ 1 to Ax + s = b, where y is the
> ```

Should this be "where y is the **vector of** binary variables"?

---

*[drake/solvers/scs_branch_and_bound.cc, line 117 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7526#-L-t4fhYj9Zqp3lQT6Td:-L-t4fhYj9Zqp3lQT6Te:b595fdc) ([raw file](https://github.com/robotlocomotion/drake/blob/f5b53283ec0988ff6b1435d6cfa0d4921ada56c5/drake/solvers/scs_branch_and_bound.cc#L117)):*
> ```cpp
>     for (; column_nonzero_index < A.p[j] - A.p[j - 1]; ++column_nonzero_index) {
>       root->A_->x[root->A_->p[j - 1] + column_nonzero_index +
>                   num_added_constraints] =
> ```

Just for readability, could you add a variable for `root->A_->p[j - 1] + column_nonzero_index +  num_added_constraints`

---

*[drake/solvers/scs_branch_and_bound.cc, line 206 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7526#-L-swNY8WVfeOoNifw3y:-L-swNY8WVfeOoNifw3z:b-d0cw0w) ([raw file](https://github.com/robotlocomotion/drake/blob/f5b53283ec0988ff6b1435d6cfa0d4921ada56c5/drake/solvers/scs_branch_and_bound.cc#L206)):*
> ```cpp
>   // We will first compute the left node, the right node will be copied and
>   // changed from the left node.
>   left_child_ = new ScsNode(A_->m - 2, A_->n - 1);
> ```

As discussed f2f, the current code leaks memory here if `Branch()` is called repeatedly.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7526#-:-L-t6XZAslSnM2BFk8le:bgh5llg)*

**Comment by RussTedrake** [comment]
Haven’t reviewed this yet, but i would have thought that the branch and bound would be written at the mathematicalprogram level (eg it dispatches to other mathematicalprogram instances for each node evaluation).  That way we can play with different solvers, etc?

**Comment by hongkai-dai** [comment]
@russtedrake I agree. I started to implement bnb for SCS, because `MathematicalProgram` object is not copyable, while I need to copy the `MathematicalProgram` from the parent node to the child node. Now I realize I can get around this copyable issue, to create a new MathematicalProgram that has the same constraints. I will submit a PR to implement bnb for general MathematicalProgram later.

---

Review status: 2 of 4 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7526#-:-L-taG9WLN1c6WYkgJEs:b-644jl0)*

**Comment by avalenzu** [comment]
@drake-jenkins-bot linux-xenial-clang-bazel-experimental-everything please

---

Reviewed 1 of 2 files at r5.
Review status: 1 of 4 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[solvers/scs_branch_and_bound.h, line 206 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/7526#-L05Mtg32chtptcrMMG4:-L05Mtg32chtptcrMMG5:b-k5a4rr) ([raw file](https://github.com/robotlocomotion/drake/blob/45ef21dee65140620795121a1c3442d8b4e2ed8f/solvers/scs_branch_and_bound.h#L206)):*
> ```objc
>   std::list<int> binary_var_indices_;
> 
>   std::shared_ptr<ScsNode> left_child_;
> ```

I already mentioned this in an email, but I'll put it here so that the record is all in one place: Do we need shared pointers here? It seems like `left_child_` and `right_child_` could be unique pointers and `parent` could be a `const` raw pointer.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7526#-:-L05h-Slr58Y1uu-ZA1N:b-rs4zsh)*

**Comment by avalenzu** [comment]
Sorry, I was on the wrong Reviewable page ... I've killed the job I just spun up.

---

Review status: 1 of 4 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7526#-:-L05hZ0BiWLOAgSQ-ma5:b-7v3rqy)*

**Comment by hongkai-dai** [comment]
Review status: 1 of 4 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[solvers/scs_branch_and_bound.h, line 206 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/7526#-L05Mtg32chtptcrMMG4:-L0AA6lrfCB6gJ1WdosJ:b65f1v3) ([raw file](https://github.com/robotlocomotion/drake/blob/45ef21dee65140620795121a1c3442d8b4e2ed8f/solvers/scs_branch_and_bound.h#L206)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

I already mentioned this in an email, but I'll put it here so that the record is all in one place: Do we need shared pointers here? It seems like `left_child_` and `right_child_` could be unique pointers and `parent` could be a `const` raw pointer.
</blockquote></details>

I used shared pointer here, because we might want to return the child nodes, and make changes on the returned child nodes. Then that returned child node pointer also owns the node. For exampe
```C++
ScsNode root();
root.Branch();
root.left_child()->Branch();

// Branching on more nodes to create a larger tree.
DoMoreBranching();

// Now I want to take one of the node in the tree, and change this node by branching with a different heuristics
std::shared_ptr<ScsNode> node = root.left_child()->left_child();
node->Branch(DifferentBranchingOption);
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7526)*

**Comment by avalenzu** [comment]
Review status: 1 of 4 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[solvers/scs_branch_and_bound.h, line 206 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/7526#-L05Mtg32chtptcrMMG4:-L0L_68nD66Qke8JPXHX:b-wi9d1r) ([raw file](https://github.com/robotlocomotion/drake/blob/45ef21dee65140620795121a1c3442d8b4e2ed8f/solvers/scs_branch_and_bound.h#L206)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

I used shared pointer here, because we might want to return the child nodes, and make changes on the returned child nodes. Then that returned child node pointer also owns the node. For exampe
```C++
ScsNode root();
root.Branch();
root.left_child()->Branch();

// Branching on more nodes to create a larger tree.
DoMoreBranching();

// Now I want to take one of the node in the tree, and change this node by branching with a different heuristics
std::shared_ptr<ScsNode> node = root.left_child()->left_child();
node->Branch(DifferentBranchingOption);
```
</blockquote></details>

This is a very strange tree at that point! :-) `node->parent()` would point to `root`, but `node` would not be a child of `root`. That seems wrong. If we're building a tree shouldn't the parent-child relationships be maintained? 

What if the child nodes were stored as a `std::vector<std::pair<ScsNode, ScsNode>>`? A new pair could be added to the vector for each call to `Branch()`. That way the tree relationships would be maintained.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7526)*

**Comment by jwnimmer-tri** [comment]
@hongkai-dai This PR seems to have stalled.  Should it still be open?

**Comment by hongkai-dai** [comment]
Close this, not useful anymore
